proposals/mutation-support-cli.md

```diff
@@ -0,0 +1,393 @@
+```
+
+**Technical Explanation**:
+
+This error message was introduced as a safeguard for when the mutation engine tries to fetch resources from an uninitialized client (the object that interfaces with a Kubernetes cluster for CRUD operations) when not in cluster mode. Without this check, the CLI would panic. See [this issue](https://github.com/kyverno/kyverno/issues/6723) for additional context.
+	
+
+## Proposal
+
+### apply
+
+The apply command will receive an additional flag, `--target-resource`, which can be passed multiple times for individual resources. For example:
+
+```bash
+kyverno apply policy-with-mutate-existing.yaml --resource=trigger1.yaml --resource=trigger2.yaml --target-resource=resource1.yaml  --target-resource=resource2.yaml 
+```
+
+Or given a directory containing resources
+
+```bash
+kyverno apply policy-with-mutate-existing.yaml /path/to/trigger/resources --target-resource=/path/to/resources 
+```
+
+When this flag is passed but the policy or policies have no mutate-existing rules, the flag should be ignored. If the flag is not passed but the policy has mutate-existing rules, a warning should be generated indicating that no resources have been passed, and the rule evaluation should be SKIPPED. The target resource flag maps to the unchanged resource, and the apply command will output the mutated resource if successful.
+
+### test
+
+An extra key will be added to the test YAML definition to define the behavior of mutate-existing with the test command:
+
+```yaml
+apiVersion: cli.kyverno.io/v1alpha1
+kind: Test
+metadata:
+  name: kyverno-test
+policies:
+  - <path/to/policy.yaml>
+  - <path/to/policy.yaml>
+resources:
+  - <path/to/resource.yaml>
+  - <path/to/resource.yaml>
+exceptions:
+  - <path/to/exception.yaml>
+  - <path/to/exception.yaml>
+targetResources: # This will be added
+  - <path/to/unmutated/resource.yaml>
+variables: variables.yaml # This will receive a new key for mutation target variables
+userinfo: user_info.yaml
+results: 
+- policy: <name> 
+  isValidatingAdmissionPolicy: false
+  rule: <name>
+  resources:
+  - <namespace_1/name_1>
+  - <namespace_2/name_2>
+  patchedResources: <file_name.yaml> # Mutated resources will be specified here, separated by ---
+  generatedResource: <file_name.yaml>
+  - <path/to/patched/resource1.yaml>
+  - <path/to/patched/resource2.yaml>
+  cloneSourceResource: <file_name.yaml>
+  kind: <kind>
+  result: pass
+```
+
+The schema of the variables file will be changed as follows:
+
+```yaml
+apiVersion: cli.kyverno.io/v1alpha1
+kind: Values
+metadata:
+  name: values
+# existing keys
+targetResources:
+  - name: nginx-demo1
+    values:
+      request.operation: CREATE
+  - name: nginx-demo2
+    values:
+      request.operation: UPDATE
+```
+
+The `targetResources` key will be added, with an array of objects containing two keys: name (a string) and values (a `map[string]interface{}`).
```

Comment:
You need an unique identifier for the target resource. Having the name only could cause the confusion if two resources are with the same name but different namespaces.

Reply:
This is how currently resources are being identified for policy variables as well.
We can improve this by adding a `namespace` variable to this schema. This will apply to target resource variables and policy variables

Reply:
The current solution needs to be improved. Though this is not a major use case, duplicate resource names could be provided to the CLI.

Reply:
Yeah, which is why i was kinda reluctant to consider this case. because it's already a pretty niche case and i didn't know if the complexity of the change will equal the value gained from addressing it.
I will include it and we can change it back later